Parse without pyparsing

AThomsen opened the pull request:
I don't expect this to be merged as it. It's more to start a discussion :-)

Is it overkill to use pyparsing for this library? The ical format seems not that complicated, but I'm probably missing something.

I've spend some hours implementing this basic parser just for fun: 
- All tests passes (except one where it expects a pyparsing exception to be thrown).
- I believe it's faster than using pyparsing.
- It handles unicode, but it's probably not as strict as the original implementation with regards to allowed characters.
- It may need to handle certain escape characters (not sure if this happens elsewhere in the library.

Fixes #469 (which I badly need for Home Assistant, as my calendars are currently broken).

allenporter reviewed (changes requested):
Thank you, appreciate the contribution, and I am definitely open to this. Pyparsing was used because the RFC specifies a grammar, so I thought it was useful to be fully conformant. I've see buggy ad-hoc implementations so it was nice to know that this is 100% correct.

Totally agree however that its overkill. I think it would be nice to keep both versions and continue running the unit tests both ways to verify they are the same. We can make an exception for the bug fix, of course.

I would like to add some additional unit tests for cases I may worry are missed by this implementation -- but otherwise agree, we want the tests should be fully exercising the spec, and if they're passing that should be sufficient. Let me double check for gaps.

allenporter reviewed (changes requested):
The parsing code has benchmarks. You can run the tests with `--benchmark-only` to show before and after improvements and add them to the PR description.

allenporter commented:
I'm adding https://github.com/allenporter/ical/pull/485 with some additional test cases. It captures two bugs (1) the original bug you reported and (2) an existing bug in parameter quoting, so we'll want to capture those fixes with this PR.

AThomsen commented:
Thanks for replying and looking into this!

I applied your comments and rebased my branch on your updates.

Also added some optimizations to the parser - as much as I could without breaking the snapshot comparison in the tests. 

Benchmark on my machine looks like this:
~~~
This implementation: `21 passed, 1176 skipped in 22.69s`
Current implementation (main): `21 passed, 1176 skipped in 24.05s`
~~~
Gains could be higher in a real-world scenario, as the new parser is thread-safe with no need for locking.

The following unit tests currently FAIL:
~~~
==================================================================== short test summary info ====================================================================
FAILED tests/parsing/test_component.py::test_invalid_contentlines[invalid_x-name-2] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/parsing/test_component.py::test_invalid_contentlines[invalid_x-name-1] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/parsing/test_component.py::test_invalid_contentlines[invalid_iana-token] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/parsing/test_component.py::test_invalid_contentlines[invalid_param_name-1] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/parsing/test_component.py::test_invalid_contentlines[invalid_param_name-2] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/test_calendar_stream.py::test_invalid_ics[control-char-value] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/test_calendar_stream.py::test_invalid_ics[control-char-name] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
FAILED tests/test_calendar_stream.py::test_invalid_ics[non-us-ascii-unicode] - Failed: DID NOT RAISE <class 'ical.exceptions.CalendarParseError'>
========================================================== 8 failed, 1168 passed, 21 skipped in 9.09s ==============================================
~~~
Besides the last one (where it's actually the test that's wrong), all the failing tests are about checking if the parsed data only uses characters according to spec. It's might be hard to implement this correctly (which is why I made this PR in the first place). 

While *writers* of ical data should totally conform to spec, is it really necessary  for *readers* to validate 100%?

allenporter commented:
"This implementation: `21 passed, 1176 skipped in 22.69s`" -- maybe double check that the skipped files are not adding a lto of overhead and only run the benchmark tests.

allenporter commented on ical/parsing/parser.py:
```diff
@@ -51,61 +35,90 @@
+                            all_values_read = char != ','
+                        pos += 1
+                    
+                dict_result[PARSE_PARAMS].append(
```
Also apply to other members of `dict_result` -- essentially drop that as a local variable.

allenporter commented on ical/parsing/component.py:
```diff
@@ -79,7 +79,6 @@ def _fold(contentline: str) -> list[str]:
         break_on_hyphens=False,
     )
 
-
```
This isn't done, still other areas not passing lnt.

allenporter commented on ical/parsing/parser.py:
```diff
@@ -54,67 +39,127 @@
-        name.set_results_name(PARSE_NAME)
-        + Group(ZeroOrMore(";" + param)).set_results_name(
-            PARSE_PARAMS, list_all_matches=True
+@cache
```
Curious what kind of performance benefits we see vs memory usage of caching this. Maybe an LRU with fixed size could be more appropriate here.

AThomsen replied:
I'll re-add the caching decorator

allenporter replied:
I'm worried about the memory usage of doing this. Whats the performance with and without? What does it look like with an LRU?

allenporter reviewed (commented):
You can use something like `pytest tests/parsing/test_component.py --benchmark-only` to view detailed benchmark results of just these tests

allenporter reviewed (commented):
While pyparing is a complex library, it seems simple to specify the grammar compared handrolliing these many loops. Unless the code can be further simplified (so far its adding 20+ lines)

If we don't see significant performance wins, then we can explore using pyparsing with a loosed character set then use the same regular expressions to validate parameter names outside of pyparsing.

Thanks for your work to improve this.

allenporter commented:
I may have found a way to do this with pyparsing here https://github.com/allenporter/ical/pull/487

AThomsen commented on ical/parsing/parser.py:
```diff
@@ -54,67 +37,104 @@
+    while True:
+        if pos >= len(line):
+            raise CalendarParseError(f"Unexpected end of line. Expected ';' or ':'", detailed_error = line)
+        char = line[pos]
+        if char == ';' or char == ':':
```
Oh, sorry. Python isn't my first language 😅

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+def parse_line(line: str) -> ParsedProperty:
+    """Parse a single line."""
+    
+    result = ParsedProperty(name="", value="", params=None)
```
Let's keep the ParsedProperty immutable and construct it at the end as the return value, similar to what I was saying about the dict return result.

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+    while True:
+        if pos >= line_len:
+            raise CalendarParseError(f"Unexpected end of line. Expected ';' or ':'", detailed_error=line)
+        if line[pos] in (';', ':'):
```
Do we loose any performance if we just get the index of ';' or ':' rather than looping? The looping here and below makes this more complex to follow so wondering if we actually see large benefits.

For example when searching through the parameter list we're doing 3 things at once needing to track the "all params ready" stuff.

Could we instead (1) search for the end of the param and (2) split by ","?

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+                        param_value_read = True
+
+                    if param_value_read:
+                        if not char in (',', ';', ':'):
```
I think this only applies when exiting the quoting so move this inside that branch?

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+            
+            while not all_values_read:
+                param_value_start = pos
+                quoted = False
+
+                if line[pos] == '"':
+                    # parameter value is quoted
+                    quoted = True
+                    pos += 1
```
We can do something like this:

```suggestion
                if quoted := (line[pos] == '"'):
                    pos += 1
```

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+                    quoted = True
+                    pos += 1
+
+                param_value_read = False
```
Lets move param_value_start down here and already bake in the +1 for quoting and remove the +1 below

allenporter commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+                    
+                    char = line[pos]
+
+                    if char == '"':
```
Can you add an additional test cases in the quoted params file with something like this?

```
NAME;PARAM-NAME="":VALUE
NAME;PARAM-NAME="",Value2,"",VALUE4,"VALUE5":VALUE
```

This catches a bug in existing pyparsing quoted string since it expects more than one character, but the rfc allows it to be empty.

allenporter reviewed (changes requested):
Yes, now that th ebug is fixed, i'm still interested in the performance improvement. Thanks for running the benchmark, that is quite a large improvement.

I have a couple more comments, but I think this is close and we can improve in further rounds after merging.

AThomsen commented on ical/parsing/parser.py:
```diff
@@ -25,96 +25,133 @@
+                pos += 1
+    else:
+        if line[pos] != ':':
+            raise CalendarParseError(f"Expected ':' after property name", detailed_error=line)
```
No, it's not hit :-)

AThomsen commented:
> > There are probably still some linting errors. I've yet to figure out how to run that locally.
> 
> I've added instructions here https://github.com/allenporter/ical/blob/main/CONTRIBUTING.md#running-pre-commit

Hmm...
```

❯ pre-commit
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/charliermarsh/ruff-pre-commit.
[INFO] Initializing environment for https://github.com/codespell-project/codespell.
[INFO] Initializing environment for https://github.com/adrienverge/yamllint.git.
[INFO] Initializing environment for https://github.com/asottile/setup-cfg-fmt.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/charliermarsh/ruff-pre-commit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/codespell-project/codespell.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/adrienverge/yamllint.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/setup-cfg-fmt.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
trim trailing whitespace.............................(no files to check)Skipped
fix end of files.....................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
check for added large files..........................(no files to check)Skipped
black................................................(no files to check)Skipped
ruff (legacy alias)..................................(no files to check)Skipped
mypy.................................................(no files to check)Skipped
codespell............................................(no files to check)Skipped
yamllint.............................................(no files to check)Skipped
setup-cfg-fmt........................................(no files to check)Skipped
```

AThomsen commented:
I followed the CONTRIBUTING guide (been using devcontainers for now), and it's now quote working as expected:
~~~
uv venv --python=3.13
Using CPython 3.13.3 interpreter at: /usr/bin/python3.13
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
~~~

This creates `.env/`, not `env/` so I had to change the `source` line to include the new "dot" path.

`pre-commit` does nothing (no files to check - see previous comment). Not sure if it's due to the env-path change.

I made some code changes as you suggested and added some tests. 

Attached are some benchmark comparisons:
* [orig-vs-new-w-caching.txt](https://github.com/user-attachments/files/20465125/orig-vs-new-w-caching.txt) - the compares the main branch (original implementation) to the new, with caching enabled.
* [new-cache-vs-nocache.txt](https://github.com/user-attachments/files/20465122/new-cache-vs-nocache.txt) - new implementation with caching vs same with no caching.

Based on these, I suggest we turn off caching. For a long-running server app like Home Assistant won't caching effectively mean that all calendar lines are kept in-memory "forever"?

As a last note, I would also suggest that you run the benchmarks yourself. I get quite varied timings on my system between test runs.

allenporter commented:
> > > There are probably still some linting errors. I've yet to figure out how to run that locally.
> > 
> > 
> > I've added instructions here https://github.com/allenporter/ical/blob/main/CONTRIBUTING.md#running-pre-commit
> 
> Hmm...
> 
> ```
> 
> ❯ pre-commit
> [INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
> [INFO] Initializing environment for https://github.com/psf/black.
> [INFO] Initializing environment for https://github.com/charliermarsh/ruff-pre-commit.
> [INFO] Initializing environment for https://github.com/codespell-project/codespell.
> [INFO] Initializing environment for https://github.com/adrienverge/yamllint.git.
> [INFO] Initializing environment for https://github.com/asottile/setup-cfg-fmt.
> [INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/psf/black.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/charliermarsh/ruff-pre-commit.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/codespell-project/codespell.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/adrienverge/yamllint.git.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/asottile/setup-cfg-fmt.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> trim trailing whitespace.............................(no files to check)Skipped
> fix end of files.....................................(no files to check)Skipped
> check yaml...........................................(no files to check)Skipped
> check for added large files..........................(no files to check)Skipped
> black................................................(no files to check)Skipped
> ruff (legacy alias)..................................(no files to check)Skipped
> mypy.................................................(no files to check)Skipped
> codespell............................................(no files to check)Skipped
> yamllint.............................................(no files to check)Skipped
> setup-cfg-fmt........................................(no files to check)Skipped
> ```

OK this is likely because the files are committed. I think you can:
(1) add some blank lines to the changed files
(2) git add <filename>
(3) re-run pre-commit

allenporter commented:
> I followed the CONTRIBUTING guide (been using devcontainers for now), and it's now quote working as expected:
> 
> ```
> uv venv --python=3.13
> Using CPython 3.13.3 interpreter at: /usr/bin/python3.13
> Creating virtual environment at: .venv
> Activate with: source .venv/bin/activate
> ```
> 
> This creates `.env/`, not `env/` so I had to change the `source` line to include the new "dot" path.
> 

Thank you, fixed the docs which missed updating this when moving to `uv`

> `pre-commit` does nothing (no files to check - see previous comment). Not sure if it's due to the env-path change.
> 
> I made some code changes as you suggested and added some tests.
> 
> Attached are some benchmark comparisons:
> 
> * [orig-vs-new-w-caching.txt](https://github.com/user-attachments/files/20465125/orig-vs-new-w-caching.txt) - the compares the main branch (original implementation) to the new, with caching enabled.
> * [new-cache-vs-nocache.txt](https://github.com/user-attachments/files/20465122/new-cache-vs-nocache.txt) - new implementation with caching vs same with no caching.
> 
> Based on these, I suggest we turn off caching. For a long-running server app like Home Assistant won't caching effectively mean that all calendar lines are kept in-memory "forever"?

Thanks, appreciate this.

> 
> As a last note, I would also suggest that you run the benchmarks yourself. I get quite varied timings on my system between test runs.

I did run these and saw anywhere from a 3x and 10x speed up -- it all seemed quite significant.

allenporter commented:
> Btw: do you know if it's possible to have a quote sign (") inside a parameter value? And what would that look like?
> 
> "Friggin \"lasers\"!" ?

I think this part of the spec is relevant:

>    Property parameter values that contain the COLON, SEMICOLON, or COMMA
>   character separators MUST be specified as quoted-string text values.
>   Property parameter values MUST NOT contain the DQUOTE character.  The
>   DQUOTE character is used as a delimiter for parameter values that
>   contain restricted characters or URI text.  For example:

>     DESCRIPTION;ALTREP="cid:part1.0001@example.org":The Fall'98 Wild
>       Wizards Conference - - Las Vegas\, NV\, USA


It looks like the purpose of the quotes is strictly to handle  COLON, SEMICOLON, or COMMA encoding only and can't be used for strings that contain quotes.

AThomsen commented on tests/parsing/__snapshots__/test_property.ambr:
```diff
@@ -48,6 +48,14 @@
+  list([
+    ParsedProperty(name='begin', value='VCALENDAR', params=None),
+    ParsedProperty(name='name', value='VALUE', params=[ParsedPropertyParameter(name='PARAM-NAME', values=['""'])]),
+    ParsedProperty(name='name', value='VALUE', params=[ParsedPropertyParameter(name='PARAM-NAME', values=['"",Value2,"",VALUE4,"VALUE5"'])]),
```
Well, this is from the `parse_basic_ics_properties` which I'm not sure why it's even there.

AThomsen commented on tests/parsing/__snapshots__/test_property.ambr:
```diff
@@ -48,6 +48,14 @@
+# name: test_from_ics[params_empty]
+  list([
+    ParsedProperty(name='begin', value='VCALENDAR', params=None),
+    ParsedProperty(name='name', value='VALUE', params=[ParsedPropertyParameter(name='PARAM-NAME', values=['""'])]),
```
Same with this one.

AThomsen commented:
Ok - pre-commit passes now :-)

I'll be gone for a few days, btw. Back sunday :-)

allenporter reviewed (approved):
Thank you very much for the contribution!

allenporter commented on ical/parsing/parser.py:
```diff
@@ -24,97 +24,153 @@
-            ) from err
+    try:
+        return [parse_line(line) for line in lines if line]
+    except Exception as err:
```
I didn't realize this was catching `Exception`.  This is catching an `IndexError` for one of the tests I added because it fails the check `if (char := line[pos]) == '"':` 

I think we're going to need to restructure this code to simplify this, taking a look.
